Reduce the amount of duplicated code in Analytics `DataStore`s

tomalec opened the pull request:
_⚠️ this PR targets [a feature branch together multiple improvements to Analytics codebase](https://github.com/woocommerce/woocommerce/pull/49425)_

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
Implements part of https://github.com/woocommerce/automatewoo/issues/1238

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

The aim of this PR is to keep the behavior and even execute the code as it was. The idea is to move the code around; the LOC reduction is not the main goal here, but rather moving the code responsible for common behavior (like caching, intervals management, etc.) to the commonplace. So, that behavior would remain consistent between specific reports, and there will be one place to fix/improve that behavior. Also, by moving shared code away, we could focus on the bahavior specific to a given report.

This is just one step that moves what we did in AutomateWoo. There is room for more improvements, as the analytics classes still contain "same but different" implementations.

- [Move caching code to shared DateStore::get_data](https://github.com/woocommerce/woocommerce/commit/304692c581523d0ea01c56ee0a9d469bed9f8626)
- [Move intervals specific initialize_queries to shared Trait](https://github.com/woocommerce/woocommerce/commit/cf9bff1054f2edd9918af0e888f033aa66d0d72d)
- [Move intervals checking logic to shared StatsDataStoreTrait](https://github.com/woocommerce/woocommerce/pull/49320/commits/19b9f75ceda4972b66f74ccdddd0b76a145ee125)
- [Use `@override` tag in DataStore PHPDocs](https://github.com/woocommerce/woocommerce/pull/49320/commits/d5f7d71b2c613d1fc6a809bd3cacc1d477af75c1)

Later, we could follow up with more changes that slightly change (unify) the behavior, like  https://github.com/woocommerce/woocommerce/compare/dev/datastore-cleanup...dev/datastore-cleanup-2?expand=1
<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Run automated tests
2. Smote test Analytics for regression

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Reduce the amount of duplicated code in Analytics `DataStore`s.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

tomalec committed:
Move caching code to shared DateStore::get_data

tomalec committed:
Move intervals specific `initialize_queries` to shared Trait

tomalec committed:
Remove the usage of DataStoreInterface,

asnow DataStore implements `get_data` method directly.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

tomalec committed:
Return data from Products DataStore

tomalec committed:
Fix whitespaces and linting errors

tomalec committed:
Merge branch 'dev/datastore-cleanup' of github.com:woocommerce/woocommerce into dev/datastore-cleanup

tomalec committed:
Replace `WPCS:` code comments with `phpcs:ignore`

tomalec committed:
Parse default query_args for Products\DataStore::get_data's include_extended_info

tomalec committed:
Move intervals checking logic to shared StatsDataStoreTrait

tomalec commented on plugins/woocommerce/src/Admin/API/Reports/Categories/DataStore.php:
```diff
@@ -201,99 +199,88 @@ protected function include_extended_info( &$categories_data, $query_args ) {
-	 * @return stdClass|WP_Error Data.
+	 * @return array Query parameters.
+	 */
+	public function get_default_query_vars() {
+		$defaults                      = parent::get_default_query_vars();
+		$defaults['category_includes'] = array();
+		$defaults['extended_info']     = false;
+
+		return $defaults;
+	}
```
:scroll: I wonder if we really need to handle default query vars here and in the Query class. Maybe we could simplify it to just one source.

tomalec commented on plugins/woocommerce/src/Admin/API/Reports/DataStore.php:
```diff
@@ -160,6 +160,54 @@ public function __construct() {
+	 *
+	 * @return array Query parameters.
+	 */
+	public function get_default_query_vars() {
+		return array(
+			'per_page' => get_option( 'posts_per_page' ),
+			'page'     => 1,
+			'order'    => 'DESC',
+			'orderby'  => 'date',
+			'before'   => TimeInterval::default_before(),
+			'after'    => TimeInterval::default_after(),
+			'fields'   => '*',
+		);
+	}
```
:scroll: I wonder if we really need to handle default query vars here and in the Query class. Maybe we could simplify it to just one source.

tomalec committed:
Use placeholder for deprecated version

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/Coupons/DataStore.php:
```diff
@@ -363,6 +250,111 @@ public static function get_coupon_id( \WC_Order_Item_Coupon $coupon_item ) {
+	 * @param array    $query_args              Query parameters.
+	 * @return stdClass|WP_Error Data object `{ totals: *, intervals: array, total: int, pages: int, page_no: int }`, or error.
+	 */
+	public function get_noncached_data( $query_args ) {
```
💅 I'm not sure if PHPDoc supports `@override` but I've seen [other repos have been using it](https://github.com/search?q=org%3Awoocommerce+%22%40override%22+language%3APHP&type=code&l=PHP), so maybe we could add `@override DataStore::get_noncached_data()`.

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/Customers/DataStore.php:
```diff
@@ -409,89 +408,16 @@ function ( $field ) {
-			$this->set_cached_data( $cache_key, $data );
-		}
+	public function get_default_query_vars() {
+		$defaults            = parent::get_default_query_vars();
+		$defaults['orderby'] = 'date_registered';
```
Before this change there were two of the default values for Customers datastore [`order_before` and `order_after`](https://github.com/woocommerce/woocommerce/pull/49320/files#diff-af075609324deed4f56396c833d794fbb0931047ed13e562851ec745c1b5c62fL429-L430). I'm not sure if those are really being used when doing the query. If yes then we should add it in the `get_default_query_vars()` method.

tomalec replied:
Thanks, I missed them taking them for just `before` and `after,` but it seems we're mapping those vars mack and worth. I'll bring them back, to keep the behavior and code as close to the original one. We could consider simplifying that in separate PRs, as those are already spanning the codebase quite wide.

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/Products/Stats/DataStore.php:
```diff
@@ -99,132 +100,126 @@ protected function update_sql_query_params( $query_args ) {
 	 */
 	public function get_data( $query_args ) {
+		// Do not include extended info like `ProductsDataStore` does.
+		return ReportsDataStore::get_data( $query_args );
```
I was thinking that `get_data()` method which defined in `Reports/DataStore.php` is not a static method, can we use `::` to call it? But I did a test locally and found it actually worked when calling a method from the parent of the parent. TIL.

tomalec replied:
:) :+1:

ianlin reviewed (commented):
Thanks for cleaning up the Data stores in analytics. Tested locally and the Analytics page worked normally. I left some comments that we can discuss.

tomalec committed:
Add @override tag to Categories DataStore

tomalec committed:
Add back order_before & _after query vars

Address https://github.com/woocommerce/woocommerce/pull/49320#discussion_r1678701590

tomalec committed:
Use `@override` tag in DataStore PHPDocs

Address https://github.com/woocommerce/woocommerce/pull/49320#discussion_r1678697159

ianlin reviewed (approved):
Thanks for the extra changes. LGTM 👍

tomalec committed:
Revert "Remove the usage of DataStoreInterface,"

This reverts commit 5b93c252fd55f148646a69ef8cc938557071bc0d.

tomalec committed:
Denote that `...\Reports\DataStore` implements `DataStoreInterface`

tomalec committed:
Bring back `get_limit_sql_params` to Orders- & Coupons/Stats/DataStore

mistakenly removed as `get_limit_params` in 19b9f75ceda4972b66f74ccdddd0b76a145ee125

tomalec committed:
Merge branch 'dev/cleanup-analytics' into dev/datastore-cleanup

tomalec commented:
I have reverted the `DataStoreInterface` removal in https://github.com/woocommerce/woocommerce/pull/49320/commits/efbf4c47ac8450254bc9b0564b29946cddb4d152 to address the suggestion from pcShBQ-21w-p2#comment-3491
> It might be a good idea to revert this change. Anyone using `DataStoreInterface` as a type hint will start receiving fatal errors. Since PHP doesn’t support duck types it isn’t enough to fit the interface, you have to explicitly declare it via `implements`.

tomalec commented:
Fixed one bug with limit params in https://github.com/woocommerce/woocommerce/pull/49320/commits/cf77942fb45c3f1ea9bac0d9163b850cf1b19c7d

I also merged the latest `trunk` -> `dev/cleanup-analytics` to hopefully help with some more flaky tests.

tomalec committed:
Replace WPCS: code comments with phpcs:ignore, fix linting errors

after the merge

tomalec committed:
Remove `order_*` from Customers/Stats/DataStore query vars

tomalec commented:
@ianlin, thank you for the review. I made some changes. Would you mind taking another look?

tomalec committed:
Move phpcs:ignore to the correct line

ianlin commented on plugins/woocommerce/changelog/49320-dev-datastore-cleanup:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: dev
+
+Reduce the amount of duplicated code in Analytics `DataStore`s. Deprecate `DataStoreInterface`.
```
Since we reverted the DataStoreInterface deprecation, we should also update the changelog.

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/Customers/Stats/DataStore.php:
```diff
@@ -60,6 +61,21 @@ protected function assign_report_columns() {
+	 * @return array Query parameters.
+	 */
+	public function get_default_query_vars() {
+		$defaults                 = ReportsDataStore::get_default_query_vars();
+		$defaults['orderby']      = 'date_registered';
```
Lint error: Equals sign not aligned.

ianlin reviewed (commented):
Thanks for the extra changes for reverting the DataStoreInterface deprecation and some bug fixes. I left some findings below, there are also some linting errors, but I think some of them could be addressed in the feature branch's PR #49425. E.g. the error `A "@since" tag was found but no version declared` for the filter.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

tomalec committed:
Fix whitespaces for PHPCS

tomalec committed:
Replace WPCS: code comments with phpcs:ignore...InterpolatedNotPrepared

tomalec committed:
Merge branch 'dev/datastore-cleanup' of github.com:woocommerce/woocommerce into dev/datastore-cleanup

ianlin reviewed (approved):
Thanks for the extra changes. LGTM.

tomalec committed:
Mention `StatsDataStoreTrait` in data.md docs

tomalec committed:
PHPCS:Ignore old `@todo` tasks as not added here.

tomalec commented:
@ianlin while [digging](https://github.com/woocommerce/woocommerce-admin/pull/2961/files) to make PHPCS pass -> find the reason behind `@todo Remove these assignements when refactoring segmenter classes to use query objects.` (currently, ignored it as not introduced here, to make GH check green) I found some developer docs https://github.com/woocommerce/woocommerce/blob/dev/datastore-cleanup/plugins/woocommerce-admin/docs/data.md

We may find it useful for our documentation phase.  I also updated it to mention the new `StatsDataStoreTrait`.

tomalec commented:
BTW, I'm always confused with [this Template check](https://github.com/woocommerce/woocommerce/actions/runs/10100193338/job/27930976911?pr=49320); I didn't change any template, and I believe none of the changes affect templates indirectly.

ianlin reviewed (approved):
Thanks for making the lint check green and the [doc](https://github.com/woocommerce/woocommerce/blob/dev/datastore-cleanup/plugins/woocommerce-admin/docs/data.md). LGTM.

> BTW, I'm always confused with [this Template check](https://github.com/woocommerce/woocommerce/actions/runs/10100193338/job/27930976911?pr=49320); I didn't change any template, and I believe none of the changes affect templates indirectly.

I've seen some other PRs had this error while they didn't change any template either. Seems to be a flaky check, maybe try to merge from trunk again before merging this PR?
